version: "3"

services:
  database:
    image: mysql:8.2.0
    container_name: database_mysql
    command: --default-authentication-plugin=caching_sha2_password
    restart: always
    volumes:
      - ./mysql:/var/lib/mysql
    environment:
      - MYSQL_DATABASE=example
      - MYSQL_ROOT_PASSWORD=example
    networks:
      - nginx_with_node_challenge
    healthcheck:
      test: ["CMD", "mysqladmin" ,"ping", "-h", "localhost", "-u", "root", "-p$$MYSQL_ROOT_PASSWORD"]
      timeout: 20s
      retries: 10
   
  node_app:
    build:
      context: .
      dockerfile: node.dockerfile
    container_name: node_app
    restart: always
    networks:
      - nginx_with_node_challenge
    depends_on:
      database:
        condition: service_healthy

  nginx:
    image: nginx:1.24.0-bullseye
    container_name: reverse_proxy
    restart: always
    volumes:
      - ./nginx.conf:/etc/nginx/nginx.conf
    ports:
      - "8080:80"
    networks:
      - nginx_with_node_challenge
    depends_on:
      - node_app
      - database

networks:
  nginx_with_node_challenge:
    driver: bridge